rewrite-java-test/src/test/java/org/openrewrite/java/JavaTemplateTest.java

```diff
@@ -428,6 +428,93 @@ void foo() {
         );
     }
 
+    @Test
+    void replaceConstructorWithStaticFactoryMethod() {
+        rewriteRun(recipeSpec -> {
+              //everything works fine when skipping type validation
+              //recipeSpec.afterTypeValidationOptions(TypeValidation.none());
+              recipeSpec.recipe(toRecipe(() -> new JavaVisitor<>() {
+                    final MethodMatcher FactoryMethod = new MethodMatcher("f.HelloFactory createHello(String, long)");
+
+                    @Override
+                    public J visitNewClass(J.NewClass newClass, ExecutionContext executionContext) {
+                        final var result = (J.NewClass) super.visitNewClass(newClass, executionContext);
+                        final Cursor parent = getCursor().getParentOrThrow();
+                        final J.MethodDeclaration methodDeclaration = parent.firstEnclosing(J.MethodDeclaration.class);
+                        final J.ClassDeclaration classDeclaration = parent.firstEnclosing(J.ClassDeclaration.class);
+                        final var isInstantiationInFactoryMethod = FactoryMethod.matches(methodDeclaration, classDeclaration);
+                        if (isInstantiationInFactoryMethod) {
+                            return result;
+                        }
+
+                        final var template = JavaTemplate.builder("createHello(#{any()}, #{any()})")
+                          .doBeforeParseTemplate(System.out::println)
+                          .staticImports("f.HelloFactory.createHello")
+                          .build();
```

Comment:
Hi! Any JavaTemplate is compiled in isolation, without any references to existing classpath entries; usually these are provided in the form of `.classpath` or `.classpathFromResources` method calls, but in other cases you can also use a minimized `.javaParser` using `.dependsOn`. In your case that would look something like this:
```suggestion
                        final var template = JavaTemplate.builder("createHello(#{any()}, #{any()})")
                          .doBeforeParseTemplate(System.out::println)
                          .javaParser(JavaParser.fromJavaVersion().dependsOn(
                            """
                              package a;
                              public class Hello {
                                public Hello(String name, long value) {}
                              }
                              """,
                            """
                              package f;
                              import a.Hello;
                              public class HelloFactory {
                                public static Hello createHello(String name, long value) {
                                  return new Hello(name, value);
                                }
                              }
                              """
                          ))
                          .staticImports("f.HelloFactory.createHello")
                          .build();
```